### PR TITLE
[TS migration] Migrate 'PDFView' component to TypeScript

### PR DESCRIPTION
**Pull request by pac-guerreiro**
### Fixed Issues

$ https://github.com/Expensify/App/issues/25151
PROPOSAL: N/A

### Tests

1. On any chat, click on the `+` sign and then select `Add attachment`
2. Select a PDF file
3. Confirm that the PDF preview renders correctly then click on send
4. On the same chat, click on the thumbnail of the attachment you just uploaded
5. Confirm that the PDF preview renders correctly

### Offline tests

Same as tests.

### QA Steps

Same as tests.

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

Screenshot_1711733842

</details>

<details>
<summary>Android: mWeb Chrome</summary>

Screenshot_1711733478

</details>

<details>
<summary>iOS: Native</summary>

Simulator Screenshot - Expensify - App - 2024-03-29 at 17 37 27

</details>

<details>
<summary>iOS: mWeb Safari</summary>

Simulator Screenshot - Expensify - Safari - 2024-03-29 at 17 30 43

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

Screenshot 2024-03-29 at 17 29 41

</details>

<details>
<summary>MacOS: Desktop</summary>

Screenshot 2024-03-29 at 17 34 31

</details>

### CODE REVIEW

File: src/components/PDFView/index.tsx [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    const [isKeyboardOpen, setIsKeyboardOpen] = useState(false);
+    const styles = useThemeStyles();
+    const {windowHeight, isSmallScreenWidth} = useWindowDimensions();
+    const prevWindowHeight = useRef(windowHeight);
+    const {translate} = useLocalize();
+
+    /**
+     * On small screens notify parent that the keyboard has opened or closed.
+     *
+     * @param {Boolean} isKeyboardOpen True if keyboard is open
```
Comment (fabioh8010) [inline src/components/PDFView/index.tsx:29]:
```suggestion
     * @param isKeyboardOpen True if keyboard is open
```

File: src/components/PDFView/index.tsx [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        }
+
+        if (!maxCanvasWidth) {
+            CanvasSize.retrieveMaxCanvasWidth();
+        }
+    };
+
+    useEffect(() => {
+        retrieveCanvasLimits();
+        // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (fabioh8010) [inline src/components/PDFView/index.tsx:61]: Add a comment explaining why you need to disable the rule

File: src/components/PDFView/index.tsx [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    }, []);
+
+    // shouldComponentUpdate(nextProps, nextState) {
+    //     return !_.isEqual(state, nextState) || !_.isEqual(props, nextProps);
+    // }
```
Comment (fabioh8010) [inline src/components/PDFView/index.tsx:66]:
```suggestion
```

Leftover?

File: src/components/PDFView/index.tsx [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+    useEffect(() => {
+        if (!isKeyboardOpen && windowHeight < prevWindowHeight.current - 100) {
```
Comment (fabioh8010) [inline src/components/PDFView/index.tsx:69]:
```suggestion
        // Use window height changes to toggle the keyboard. To maintain keyboard state
        // on all platforms we also use focus/blur events. So we need to make sure here
        // that we avoid redundant keyboard toggling.
        // Minus 100px is needed to make sure that when the internet connection is
        // disabled in android chrome and a small 'No internet connection' text box appears,
        // we do not take it as a sign to open the keyboard
        if (!isKeyboardOpen && windowHeight < prevWindowHeight.current - 100) {
```

File: src/components/PDFView/index.tsx [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+import CONST from '@src/CONST';
+import ONYXKEYS from '@src/ONYXKEYS';
+import PDFPasswordForm from './PDFPasswordForm';
+import type {PDFViewOnyxProps, PDFViewProps} from './types';
+
+function PDFView({onToggleKeyboard, fileName, onPress, isFocused, sourceURL, errorLabelStyles, maxCanvasArea, maxCanvasHeight, maxCanvasWidth, style}: PDFViewProps) {
+    const [isKeyboardOpen, setIsKeyboardOpen] = useState(false);
+    const styles = useThemeStyles();
+    const {windowHeight, isSmallScreenWidth} = useWindowDimensions();
+    const prevWindowHeight = useRef(windowHeight);
```
Comment (fabioh8010) [inline src/components/PDFView/index.tsx:23]:
```suggestion
    const prevWindowHeight = usePrevious(windowHeight);
```

File: src/components/PDFView/index.tsx [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    const renderPDFView = () => {
+        const outerContainerStyle = [styles.w100, styles.h100, styles.justifyContentCenter, styles.alignItemsCenter];
+
+        return (
+            <View
+                style={outerContainerStyle}
+                tabIndex={0}
+            >
+                <PDFPreviewer
+                    contentContainerStyle={style}
```
Comment (fabioh8010) [inline src/components/PDFView/index.tsx:85]:
```suggestion
                    contentContainerStyle={style as CSSProperties}
```

Let's just assert here so we don't have to change the other places to use `CSSProperties`

File: src/components/PDFView/types.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    fileName?: string;
+
+    /** (web only) Additional style props */
+    style?: CSSProperties;
+
+    /** Notify parent that the keyboard has opened or closed */
+    onToggleKeyboard?: (isKeyboardOpen: boolean) => void;
+
+    /** Handles press events like toggling attachment arrows natively */
+    onPress?: (event?: GestureResponderEvent | KeyboardEvent | undefined) => void | Promise<void>;
```
Comment (fabioh8010) [inline src/components/PDFView/types.ts:20]:
```suggestion
    onPress?: (event?: GestureResponderEvent | KeyboardEvent) => void;
```

File: src/components/PDFView/types.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    onScaleChanged?: (newScale: number) => void;
+
+    /** Handles load complete event in PDF component */
+    onLoadComplete: (path: string) => void;
+
+    /** Should focus to the password input  */
+    isFocused?: boolean;
+
+    /** Styles for the error label */
+    errorLabelStyles: StyleProp<TextStyle>;
```
Comment (fabioh8010) [inline src/components/PDFView/types.ts:32]:
```suggestion
    errorLabelStyles?: StyleProp<TextStyle>;
```

File: src/components/PDFView/types.ts [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+};
+
+type PDFViewOnyxProps = {
+    maxCanvasArea: OnyxEntry<number>;
+    maxCanvasHeight: OnyxEntry<number>;
+    maxCanvasWidth: OnyxEntry<number>;
```
Comment (fabioh8010) [inline src/components/PDFView/types.ts:38]: Let's add description for each prop

File: src/components/PDFView/PDFInfoMessage.tsx [hunk 10]
```diff
@@ -1,25 +1,24 @@
...
 import useThemeStyles from '@hooks/useThemeStyles';
 import variables from '@styles/variables';
 
-const propTypes = {
+type PDFInfoMessageProps = {
     /** Callback function to indicate that PDF password form should be shown */
-    onShowForm: PropTypes.func.isRequired,
-
-    ...withLocalizePropTypes,
+    onShowForm: () => void | ((event: GestureResponderEvent) => void);
```
Comment (fabioh8010) [inline src/components/PDFView/PDFInfoMessage.tsx:15]:
```suggestion
    onShowForm: (event: GestureResponderEvent | KeyboardEvent) => void;
```

Import `KeyboardEvent` from `react`

File: src/components/TextLink.tsx
Comment (fabioh8010) [inline src/components/TextLink.tsx:1]: Let's apply this changes on `TextLink`

```diff
diff --git a/src/components/TextLink.tsx b/src/components/TextLink.tsx
index 8b44b2edbe..b6ffb14753 100644
--- a/src/components/TextLink.tsx
+++ b/src/components/TextLink.tsx
@@ -1,4 +1,4 @@
-import type {ForwardedRef, KeyboardEventHandler, MouseEventHandler} from 'react';
+import type {ForwardedRef, KeyboardEvent, KeyboardEventHandler, MouseEventHandler} from 'react';
 import React, {forwardRef} from 'react';
 // eslint-disable-next-line no-restricted-imports
 import type {GestureResponderEvent, Text as RNText, StyleProp, TextStyle} from 'react-native';
@@ -20,7 +20,7 @@ type PressProps = {
     href?: undefined;
 
     /** Overwrites the default link behavior with a custom callback */
-    onPress: () => void | ((event: GestureResponderEvent) => void);
+    onPress: (event: GestureResponderEvent | KeyboardEvent) => void;
 };
 
 type TextLinkProps = (LinkProps | PressProps) &
@@ -36,9 +36,9 @@ function TextLink({href, onPress, children, style, onMouseDown = (event) => even
     const {environmentURL} = useEnvironment();
     const styles = useThemeStyles();
 
-    const openLink = () => {
+    const openLink = (event: GestureResponderEvent | KeyboardEvent) => {
         if (onPress) {
-            onPress();
+            onPress(event);
         } else {
             Link.openLink(href, environmentURL);
         }
@@ -47,7 +47,7 @@ function TextLink({href, onPress, children, style, onMouseDown = (event) => even
     const openLinkOnTap = (event: GestureResponderEvent) => {
         event.preventDefault();
 
-        openLink();
+        openLink(event);
     };
 
     const openLinkOnEnterKey: KeyboardEventHandler = (event) => {
@@ -56,7 +56,7 @@ function TextLink({href, onPress, children, style, onMouseDown = (event) => even
         }
         event.preventDefault();
 
-        openLink();
+        openLink(event);
     };
 
     return (

```

File: src/components/PDFView/index.tsx (hunk 3 above)
Comment (pac-guerreiro) [reply src/components/PDFView/index.tsx:66]: Yes 😅

File: src/components/PDFView/index.tsx (hunk 4 above)
Comment (pac-guerreiro) [reply src/components/PDFView/index.tsx:69]: @fabioh8010 you were inspired when you wrote this comment, me gusta 😄 

Thank you for the info!

**Comment by pac-guerreiro** [comment]
@fabioh8010 @kubabutkiewicz @ishpaul777 

All feedback addressed, tests added, author checklist complete and screenshots added!

**Review by kubabutkiewicz — APPROVED** [review]

File: src/components/PDFView/index.tsx [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+                            isPasswordInvalid={isPasswordInvalid}
+                            onSubmit={onSubmit}
+                            onPasswordUpdated={onPasswordChange}
+                            onPasswordFieldFocused={toggleKeyboardOnSmallScreens}
+                        />
+                    )}
+                />
+            </View>
+        );
+    };
```
Comment (blazejkustra) [inline src/components/PDFView/index.tsx:112]: Shouldn't this be a separate component? This looks like a big calculation each time PDFView rerenders 😄

**Review by blazejkustra — APPROVED** [review]

File: src/components/PDFView/index.tsx (hunk 11 above)
Comment (fabioh8010) [reply src/components/PDFView/index.tsx:112]: I would avoid this kind of refactor right now, let's focus on the migration only

**Comment by pac-guerreiro** [comment]
@roryabraham @mountiny @cubuspl42 

@ishpaul777 is the C+ reviewing this PR 😄

**Comment by mountiny** [comment]
Assigned @roryabraham do you want to handle this one or should I go ahead?

**Comment by fabioh8010** [comment]
> @pac-guerreiro Can you please merge main

@pac-guerreiro bump

**Review by ishpaul777 — APPROVED** [review]
LGTM!

**Comment by pac-guerreiro** [comment]
@ishpaul777 @fabioh8010 just merged main into this branch! sorry for the delay!

**Comment by blazejkustra** [comment]
@ishpaul777 Can you approve again so an internal is assigned? Thanks!

**Review by ishpaul777 — APPROVED** [review]

**Review by AndrewGable — APPROVED** [review]

**Pull request merged**